shuffle-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/ucx/UCX.scala

```diff
@@ -425,36 +300,57 @@ class UCX(transport: UCXShuffleTransport, executor: BlockManagerId, rapidsConf:
         s"Invalid Active Message re-registration of response handler for " +
           s"${UCX.formatAmIdAndHeader(activeMessageId, header)}")
     }
+
+    def removeResponseActiveMessageHandler(header: Long): Unit = {
+      responseCallbacks.remove(header)
+    }
   }
 
   /**
    * Register a response handler (clients will use this)
    *
    * @note This function will be called for each client, with the same `am.activeMessageId`
-   * @param activeMessageId (up to 5 bits) used to register with UCX an Active Message
-   * @param header a long used to demux responses arriving at `activeMessageId`
+   * @param am UCXActiveMessage (id and header) to register a one time response for
    * @param responseCallback callback to handle a particular response
    */
-  def registerResponseHandler(
-      activeMessageId: Int, header: Long, responseCallback: UCXAmCallback): Unit = {
-    logDebug(s"Register Active Message " +
-      s"${UCX.formatAmIdAndHeader(activeMessageId, header)} response handler")
+  def registerResponseHandler(am: UCXActiveMessage, responseCallback: UCXAmCallback): Unit = {
+    logDebug(s"Register Active Message $am response handler") 
 
-    amRegistrations.computeIfAbsent(activeMessageId,
+    amRegistrations.computeIfAbsent(am.activeMessageId,
       _ => {
-        val reg = new ResponseActiveMessageRegistration(activeMessageId)
+        val reg = new ResponseActiveMessageRegistration(am.activeMessageId)
         registerActiveMessage(reg)
         reg
       }) match {
       case reg: ResponseActiveMessageRegistration =>
-        reg.addResponseActiveMessageHandler(header, responseCallback)
+        reg.addResponseActiveMessageHandler(am.header, responseCallback)
       case other =>
         throw new IllegalStateException(
-          s"Attempted to add a response Active Message handler to existing registration $other " +
-            s"for ${UCX.formatAmIdAndHeader(activeMessageId, header)}")
+          s"Attempted to add a response Active Message $am handler but got $other")
     }
   }
 
+  /**
+   * Unregister an active message handler (given by `am.header`). This is invoked
+   * from request error handlers. On request failure (which is terminal for the transaction),
+   * the response will never arrive, so this cleans up the response handler.
+   *
+   * @param am active message (id, header) that for the response handler to remove
+   */
+  def unregisterResponseHandler(am: UCXActiveMessage): Unit = {
+    amRegistrations.computeIfPresent(am.activeMessageId,
+      (_, msg) => {
+        msg match {
+          case reg: ResponseActiveMessageRegistration =>
+            reg.removeResponseActiveMessageHandler(am.header)
```

Comment:
Do we need to do anything special if this was the last handler removed?

Reply:
Yeah, as long as the map can't grow unbounded with empty handler stubs then it should be fine as-is.